### PR TITLE
[DX-2248] Port grind Key fix 

### PR DESCRIPTION
**Pull request by carleeto**
### Summary
This is covering all the backwards compatible scenarios, with versions of core-sdk

Between 1.0.0-beta.1 and 2.0.0
version 2.0.1
JS/Link SDK

#### Why the changes
https://immutable.atlassian.net/browse/DX-2167
https://immutable.atlassian.net/browse/DX-2184

### Things worth calling out
See https://github.com/immutable/imx-core-sdk/pull/341

https://github.com/immutable/imx-sdk-js/pull/361
https://github.com/immutable/imx-core-sdk-golang/pull/127

### CODE REVIEW

File: imx/signers/stark/factory.go [hunk 1]
```diff
@@ -66,9 +71,85 @@ func GenerateLegacyKey(signer imx.L1Signer) (string, error) {
...
+	if err != nil {
+		return "", fmt.Errorf("creating stark signer: %w", err)
+	}
+	// If the user account matches with generated stark public key user, just return Stark Private Key.
+	if registeredStarkPublicKey == starkSigner.GetPublicKey() {
+		return starkPrivateKey, nil
+	}
+
+	// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+	// This is backwards compatible crypto (core-sdk) version 2.0.1
```
Comment (carleeto) [inline imx/signers/stark/factory.go:123]: need to fix this 
```suggestion
	// This is backwards compatible
```

**Review by suluvai — APPROVED** [review]

File: imx/signers/stark/factory.go [hunk 2]
```diff
@@ -145,3 +282,32 @@ func getStarkPath(layerName, applicationName, ethereumAddress, index string) str
...
+	}
+	// err is nil from here on
+	if response == nil {
+		return "", false, imx.NewIMXError(httpResponse, errors.New("nil response"))
+	}
+	accounts := response.GetAccounts()
+	if len(accounts) == 0 {
+		return "", false, imx.NewIMXError(httpResponse, errors.New("successful API call, but 0 accounts returned"))
+	}
+	return accounts[0], false, nil
```
Comment (igx-C) [inline imx/signers/stark/factory.go:312]: To cover all invariants we could check for len(accounts) != 1, it shouldn't exist

File: imx/signers/stark/factory_test.go [hunk 3]
```diff
@@ -19,12 +19,35 @@ func TestStarkSignerFactory_GenerateRandomKey(t *testing.T) {
...
-
-	assert.Equalf(t, expectedKeyValueAfterGrinding, grind(privateKey), "Verify grinding logic")
+	type test struct {
+		name                 string
+		privateKey           string
+		wantKeyAfterGrinding string
+	}
+	tests := []test{
+		{
+			name:                 "correct ground key",
```
Comment (igx-C) [inline imx/signers/stark/factory_test.go:29]:
```suggestion
			name:                 "correct grind key",
```

File: imx/signers/stark/factory.go (hunk 2 above)
Comment (suluvai) [reply imx/signers/stark/factory.go:312]: Good point @igx-C,

I would prefer to keep it as is, to keep it consistent with the released `core-sdk` and `sdk-js` versions

Comment (igx-C) [reply imx/signers/stark/factory.go:312]: 👍

**Review by igx-C — APPROVED** [review]

**Review by cajames — APPROVED** [review]

**Pull request merged**